File: backend/application-service/src/models/application.py
"""
Application Domain Model
"""
from sqlalchemy import Column, String, Text, Integer, DateTime, Enum
from sqlalchemy.dialects.postgresql import UUID
from datetime import datetime
import uuid
import enum

import sys
sys.path.append('../../shared')
from database import Base


class ApplicationStatus(str, enum.Enum):
    PENDING = "pending"
    REVIEWED = "reviewed"
    INVITED = "invited"
    REJECTED = "rejected"


class Application(Base):
    """Job Application entity"""
    __tablename__ = "applications"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    job_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    employee_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    cv_id = Column(UUID(as_uuid=True), nullable=True)
    status = Column(Enum(ApplicationStatus), nullable=False, default=ApplicationStatus.PENDING)
    match_score = Column(Integer, nullable=True)  # 0-100
    match_summary = Column(Text, nullable=True)
    applied_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    def to_dict(self):
        return {
            'id': str(self.id),
            'job_id': str(self.job_id),
            'employee_id': str(self.employee_id),
            'cv_id': str(self.cv_id) if self.cv_id else None,
            'status': self.status.value if isinstance(self.status, ApplicationStatus) else self.status,
            'match_score': self.match_score,
            'match_summary': self.match_summary,
            'applied_at': self.applied_at.isoformat() if self.applied_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }

File: frontend/talentlink_frontend/lib/models/application.dart
class Application {
  final String id;
  final String jobId;
  final String employeeId;
  final String? cvId;
  final String status;
  final double? matchScore;
  final String? matchSummary;
  final DateTime appliedAt;
  final Map<String, dynamic>? matchDetails;

  Application({
    required this.id,
    required this.jobId,
    required this.employeeId,
    this.cvId,
    required this.status,
    this.matchScore,
    this.matchSummary,
    required this.appliedAt,
    this.matchDetails,
  });

  factory Application.fromJson(Map<String, dynamic> json) {
    return Application(
      id: json['id'] ?? '',
      jobId: json['job_id'] ?? '',
      employeeId: json['employee_id'] ?? '',
      cvId: json['cv_id'],
      status: json['status'] ?? 'pending',
      matchScore: json['match_score']?.toDouble(),
      matchSummary: json['match_summary'],
      appliedAt: DateTime.parse(json['applied_at'] ?? DateTime.now().toIso8601String()),
      matchDetails: json['match_details'],
    );
  }

  String getStatusDisplay() {
    switch (status.toLowerCase()) {
      case 'pending':
        return 'Pending Review';
      case 'reviewed':
        return 'Reviewed';
      case 'invited':
        return 'Interview Invited';
      case 'rejected':
        return 'Not Selected';
      default:
        return status;
    }
  }

  String getMatchScoreDisplay() {
    if (matchScore == null) return 'Calculating...';
    return '${matchScore!.toStringAsFixed(0)}%';
  }
}
